be/src/vec/common/sort/sorter.cpp

```diff
@@ -17,48 +17,154 @@
 
 #include "vec/common/sort/sorter.h"
 
+#include "runtime/block_spill_manager.h"
 #include "runtime/thread_context.h"
 
 namespace doris::vectorized {
 
-void MergeSorterState::build_merge_tree(SortDescription& sort_description) {
-    for (const auto& block : sorted_blocks) {
-        cursors.emplace_back(block, sort_description);
+// When doing spillable sorting, each sorted block is spilled into a single file.
+//
+// In order to decrease memory pressure when merging
+// multiple spilled blocks into one bigger sorted block, only part
+// of each spilled blocks are read back into memory at a time.
+//
+// Currently the spilled blocks are splitted into small sub blocks,
+// each sub block is serialized in PBlock format and appended
+// to the spill file.
+//
+// This number specifies the maximum size of sub blocks
+static constexpr int BLOCK_SPILL_BATCH_BYTES = 8 * 1024 * 1024;
+
+Status MergeSorterState::add_sorted_block(Block& block) {
+    auto rows = block.rows();
+    if (0 == rows) {
+        return Status::OK();
+    }
+    if (0 == avg_row_bytes_) {
+        avg_row_bytes_ = block.bytes() / rows;
+        spill_block_batch_size_ = (BLOCK_SPILL_BATCH_BYTES + avg_row_bytes_ - 1) / avg_row_bytes_;
     }
 
-    if (sorted_blocks.size() > 1) {
-        for (auto& cursor : cursors) priority_queue.push(MergeSortCursor(&cursor));
+    auto bytes_used = data_size();
+    if (is_spilled_ ||
+        (config::external_sort_bytes_threshold > 0 &&
+         (bytes_used + block.allocated_bytes()) >=
+                 (config::external_sort_bytes_threshold - BLOCK_SPILL_BATCH_BYTES))) {
+        is_spilled_ = true;
+        BlockSpillWriterUPtr spill_block_wirter;
+        RETURN_IF_ERROR(ExecEnv::GetInstance()->block_spill_mgr()->get_writer(
+                spill_block_batch_size_, spill_block_wirter));
+
+        RETURN_IF_ERROR(spill_block_wirter->write(block));
+        spilled_sorted_block_streams_.emplace_back(spill_block_wirter->get_id());
+
+        spill_block_wirter->close();
+
+        if (init_merge_sorted_block_) {
+            init_merge_sorted_block_ = false;
+            merge_sorted_block_ = block.clone_empty();
+        }
+    } else {
+        sorted_blocks_.emplace_back(std::move(block));
     }
+    num_rows_ += rows;
+    return Status::OK();
+}
+
+Status MergeSorterState::_build_merge_tree_not_spilled(const SortDescription& sort_description) {
+    for (const auto& block : sorted_blocks_) {
+        cursors_.emplace_back(block, sort_description);
+    }
+
+    if (sorted_blocks_.size() > 1) {
+        for (auto& cursor : cursors_) priority_queue_.push(MergeSortCursor(&cursor));
+    }
+    return Status::OK();
+}
+
+Status MergeSorterState::build_merge_tree(const SortDescription& sort_description) {
```

Comment:
Status TopNSorter::prepare_for_read() {
    _state->build_merge_tree(_sort_description);
    return Status::OK();

Please check all usage of build_merge_tree and ensure that they will check the return status. I find they do not check.